rust/loqui_connection/src/supervisor.rs

```diff
@@ -0,0 +1,146 @@
+use crate::async_backoff::AsyncBackoff;
+use crate::connection::Connection;
+use crate::error::{convert_timeout_error, LoquiError};
+use crate::handler::Handler;
+use failure::Error;
+use futures::sync::mpsc::{self, Receiver, Sender, UnboundedReceiver, UnboundedSender};
+use futures::sync::oneshot;
+use futures_timer::FutureExt;
+use std::net::SocketAddr;
+use std::time::Instant;
+use tokio::await;
+use tokio::net::TcpStream;
+use tokio::prelude::*;
+
+/// A connection supervisor. It will keep the connection alive until its senders are dropped.
+/// Supports backoff.
+pub struct Supervisor<H: Handler> {
+    event_sender: Sender<H::InternalEvent>,
+    /// A Sender that will drop once the client is dropped. If it's dropped then we should
+    /// stop trying to connect.
+    _close_sender: UnboundedSender<()>,
+}
+
+impl<H: Handler> Supervisor<H> {
+    /// Spawns a new supervisor. Waits until the connection is ready before returning.
+    ///
+    /// # Arguments
+    ///
+    /// * `address` - The address to connect to
+    /// * `handler_creator` - a `Fn` that creates a `Handler`. Called each time a new TCP connection is made.
+    /// * `queue_size` - the number of requests the supervisor should hold before timing out requests.
+    pub async fn connect<C>(
+        address: SocketAddr,
+        handler_creator: C,
+        queue_size: usize,
+    ) -> Result<Self, Error>
+    where
+        C: Fn() -> H + Send + 'static,
+    {
+        let (event_sender, event_rx) = mpsc::channel(queue_size);
+        let (_close_sender, close_rx) = mpsc::unbounded::<()>();
+        let connection = Self {
+            event_sender,
+            _close_sender,
+        };
+        let (ready_tx, ready_rx) = oneshot::channel();
+        tokio::spawn_async(
+            async move {
+                if let Err(e) = await!(run(address, handler_creator, event_rx, close_rx, ready_tx))
+                {
+                    warn!("Supervisor error. error={:?}", e);
+                }
+            },
+        );
+        await!(ready_rx)?;
+        debug!("Supervisor ready.");
+        Ok(connection)
+    }
+
+    pub async fn send(&self, event: H::InternalEvent, deadline: Instant) -> Result<(), Error> {
+        let send_with_deadline = self
+            .event_sender
+            .clone()
```

Comment:
Hmm, this is the error if you don't clone:
```
error[E0507]: cannot move out of borrowed content
  --> loqui_connection/src/supervisor.rs:61:34
   |
61 |           let send_with_deadline = self
   |  __________________________________^
62 | |             .event_sender
   | |_________________________^ cannot move out of borrowed content

error: aborting due to previous error
```

`send` moves the sender. https://docs.rs/futures/0.1/futures/sink/trait.Sink.html#method.send

Reply:
I don't know if this matters anymore since the supervisor was removed, but I think the idea is that you call send and then use the sender it returns for future sends -- I wonder what the perf impact of cloning every time vs. wrapping it in an `Option` and `take`ing it every time you use it.

Reply:
I think that would mean `&self` would need to change to `&mut self`, right?